api/v1beta1/frrconfiguration_types.go

```diff
@@ -143,6 +143,12 @@ type Neighbor struct {
 	// +optional
 	BFDProfile string `json:"bfdProfile,omitempty"`
 
+	// EnableGracefulRestart allows BGP peer to continue to forward data packets along
```

Comment:
yeah, I want to discuss that with you.
I tried but when I deploy `make deploy-controller` I get this error
```
The CustomResourceDefinition "frrconfigurations.frrk8s.metallb.io" is invalid: spec.validation.openAPIV3Schema.properties[spec].properties[bgp].properties[routers].items.properties[neighbors].items.properties[enableGracefulRestart].x-kubernetes-validations[0].rule: Invalid value: "self == oldSelf": oldSelf cannot be used on the uncorrelatable portion of the schema within spec.validation.openAPIV3Schema.properties[spec].properties[bgp].properties[routers]
```
which looks we can not use this in `uncorrelatable portion of the schema`. Maybe I am doing something wrong not sure

https://github.com/kubernetes/apiextensions-apiserver/blob/master/pkg/apis/apiextensions/validation/validation.go#L852

Not sure what is the way forward (if any)
1. custom logic to execute a reset on the bgp neighbor if this flag is added
2. custom logic to just show degraded status/warning
3. any other good idea?

Reply:
> yeah, I want to discuss that with you. I tried but when I deploy `make deploy-controller` I get this error
> 
> ```
> The CustomResourceDefinition "frrconfigurations.frrk8s.metallb.io" is invalid: spec.validation.openAPIV3Schema.properties[spec].properties[bgp].properties[routers].items.properties[neighbors].items.properties[enableGracefulRestart].x-kubernetes-validations[0].rule: Invalid value: "self == oldSelf": oldSelf cannot be used on the uncorrelatable portion of the schema within spec.validation.openAPIV3Schema.properties[spec].properties[bgp].properties[routers]
> ```
> 
> which looks we can not use this in `uncorrelatable portion of the schema`. Maybe I am doing something wrong not sure
> 
> https://github.com/kubernetes/apiextensions-apiserver/blob/master/pkg/apis/apiextensions/validation/validation.go#L852
> 
> Not sure what is the way forward (if any)
> 
>     1. custom logic to execute a reset on the bgp neighbor if this flag is added
> 

This is bad, as the user would declare a desired state that not only we can't fulfill, but we also change without him knowing.

>     2. custom logic to just show degraded status/warning
> 

We could have the validation done both in the webhook / config loading, but it's not ideal

>     3. any other good idea?

Looking at the source code!

The error comes from here

https://github.com/kubernetes/apiextensions-apiserver/blob/5d129eee2b85758751c8a23f64ff2d647dd48916/pkg/apis/apiextensions/validation/validation.go#L1261

It fails if 

```
if uncorrelatablePath := ssv.forbidOldSelfValidations(); uncorrelatablePath != nil {
```

which comes from here https://github.com/kubernetes/apiextensions-apiserver/blob/5d129eee2b85758751c8a23f64ff2d647dd48916/pkg/apis/apiextensions/validation/validation.go#L1397:

which is being set on 

https://github.com/kubernetes/apiextensions-apiserver/blob/5d129eee2b85758751c8a23f64ff2d647dd48916/pkg/apis/apiextensions/validation/validation.go#L1075


```
		// we can only correlate old/new items for "map" and "set" lists, and correlation of
		// "set" elements by identity is not supported for cel (x-kubernetes-validations)
		// rules. an unset list type defaults to "atomic".
```

It basically means, to make assumptions on the old value we must be able to understand who is what.

I made it work by:

- declaring the neighbor list as map, with the address as key
- declaring the router list as a map, with vrf as key


Diff as follows:

```
diff --git a/api/v1beta1/frrconfiguration_types.go b/api/v1beta1/frrconfiguration_types.go
index 430e88e..8ec9c32 100644
--- a/api/v1beta1/frrconfiguration_types.go
+++ b/api/v1beta1/frrconfiguration_types.go
@@ -56,6 +56,8 @@ type RawConfig struct {
 type BGPConfig struct {
        // Routers is the list of routers we want FRR to configure (one per VRF).
        // +optional
+       // +listType=map
+       // +listMapKey=vrf
        Routers []Router `json:"routers"`
        // BFDProfiles is the list of bfd profiles to be used when configuring the neighbors.
        // +optional
@@ -73,8 +75,11 @@ type Router struct {
        ID string `json:"id,omitempty"`
        // VRF is the host vrf used to establish sessions from this router.
        // +optional
-       VRF string `json:"vrf,omitempty"`
+       // +kubebuilder:default:=""
+       VRF string `json:"vrf"`
        // Neighbors is the list of neighbors we want to establish BGP sessions with.
+       // +listType=map
+       // +listMapKey=address
        // +optional
        Neighbors []Neighbor `json:"neighbors,omitempty"`
        // Prefixes is the list of prefixes we want to advertise from this router instance.
@@ -82,6 +87,7 @@ type Router struct {
        Prefixes []string `json:"prefixes,omitempty"`
 }

+// +kubebuilder:validation:XValidation:rule="!has(oldSelf.enableGracefulRestart) || has(self.enableGracefulRestart)", message="Value is required once set"
 // Neighbor represents a BGP Neighbor we want FRR to connect to.
 type Neighbor struct {
        // ASN is the AS number to use for the local end of the session.
@@ -96,6 +102,7 @@ type Neighbor struct {
        SourceAddress string `json:"sourceaddress,omitempty"`

        // Address is the IP address to establish the session with.
+       //+kubebuilder:validation:Required
        Address string `json:"address"`

        // Port is the port to dial when establishing the session.
@@ -146,6 +153,7 @@ type Neighbor struct {
        // EnableGracefulRestart allows BGP peer to continue to forward data packets along
        // known routes while the routing protocol information is being restored.
        // This field is immutable because it requires restart of the BGP session
+       // +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Value is immutable"
        // +optional
        EnableGracefulRestart bool `json:"enableGracefulRestart,omitempty"`
```